File: src/yazses/ipc/protocol.py
"""Minimal JSON-RPC 2.0 framing.

Wire format: newline-delimited JSON. Each request and each response is a single
UTF-8 line terminated with '\\n'. No notifications, no batching. Errors follow
JSON-RPC 2.0 conventions but with YazSes-specific codes in the -32099..-32000
"server error" range.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any

JSONRPC_VERSION = "2.0"

# Standard JSON-RPC error codes
PARSE_ERROR = -32700
INVALID_REQUEST = -32600
METHOD_NOT_FOUND = -32601
INVALID_PARAMS = -32602
INTERNAL_ERROR = -32603

# YazSes-specific
NOT_REACHABLE = -32001  # client side — daemon socket not listening
HANDLER_FAILED = -32002  # server side — handler raised


@dataclass
class Request:
    method: str
    params: dict[str, Any] = field(default_factory=dict)
    id: int | str | None = None

    def to_json(self) -> str:
        return json.dumps(
            {
                "jsonrpc": JSONRPC_VERSION,
                "method": self.method,
                "params": self.params,
                "id": self.id,
            },
            separators=(",", ":"),
        )

    @classmethod
    def from_json(cls, line: str) -> Request:
        data = json.loads(line)
        if data.get("jsonrpc") != JSONRPC_VERSION:
            raise ValueError(f"Unexpected jsonrpc version: {data.get('jsonrpc')!r}")
        method = data.get("method")
        if not isinstance(method, str):
            raise ValueError("Request missing 'method' string")
        params = data.get("params") or {}
        if not isinstance(params, dict):
            raise ValueError("Request 'params' must be an object")
        return cls(method=method, params=params, id=data.get("id"))


@dataclass
class RpcError:
    code: int
    message: str
    data: Any = None

    def to_dict(self) -> dict[str, Any]:
        out: dict[str, Any] = {"code": self.code, "message": self.message}
        if self.data is not None:
            out["data"] = self.data
        return out


@dataclass
class Response:
    id: int | str | None
    result: Any = None
    error: RpcError | None = None

    def to_json(self) -> str:
        body: dict[str, Any] = {"jsonrpc": JSONRPC_VERSION, "id": self.id}
        if self.error is not None:
            body["error"] = self.error.to_dict()
        else:
            body["result"] = self.result
        return json.dumps(body, separators=(",", ":"))

    @classmethod
    def from_json(cls, line: str) -> Response:
        data = json.loads(line)
        if data.get("jsonrpc") != JSONRPC_VERSION:
            raise ValueError(f"Unexpected jsonrpc version: {data.get('jsonrpc')!r}")
        err = data.get("error")
        rpc_error: RpcError | None = None
        if err is not None:
            rpc_error = RpcError(
                code=int(err.get("code", INTERNAL_ERROR)),
                message=str(err.get("message", "")),
                data=err.get("data"),
            )
        return cls(id=data.get("id"), result=data.get("result"), error=rpc_error)

File: crates/yazses-ipc/src/protocol.rs
// JSON-RPC 2.0 framing — mirrors v0.4 yazses/ipc/protocol.py exactly.
// One connection per request; newline-delimited JSON; no batching.

use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const JSONRPC_VERSION: &str = "2.0";

// JSON-RPC error codes
pub const PARSE_ERROR: i32 = -32700;
pub const INVALID_REQUEST: i32 = -32600;
pub const METHOD_NOT_FOUND: i32 = -32601;
pub const INVALID_PARAMS: i32 = -32602;
pub const INTERNAL_ERROR: i32 = -32603;
pub const HANDLER_FAILED: i32 = -32000;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Request {
    pub jsonrpc: String,
    pub method: String,
    #[serde(default)]
    pub params: serde_json::Map<String, Value>,
    pub id: Option<Value>,
}

impl Request {
    pub fn parse(line: &str) -> anyhow::Result<Self> {
        let req: Self = serde_json::from_str(line)?;
        anyhow::ensure!(
            req.jsonrpc == JSONRPC_VERSION,
            "unsupported jsonrpc version: {}",
            req.jsonrpc
        );
        Ok(req)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RpcError {
    pub code: i32,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub data: Option<Value>,
}

impl RpcError {
    pub fn new(code: i32, message: impl Into<String>) -> Self {
        Self {
            code,
            message: message.into(),
            data: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Response {
    pub jsonrpc: String,
    pub id: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub result: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<RpcError>,
}

impl Response {
    pub fn ok(id: Option<Value>, result: impl Serialize) -> anyhow::Result<Self> {
        Ok(Self {
            jsonrpc: JSONRPC_VERSION.into(),
            id,
            result: Some(serde_json::to_value(result)?),
            error: None,
        })
    }

    pub fn err(id: Option<Value>, error: RpcError) -> Self {
        Self {
            jsonrpc: JSONRPC_VERSION.into(),
            id,
            result: None,
            error: Some(error),
        }
    }

    pub fn to_line(&self) -> anyhow::Result<String> {
        Ok(serde_json::to_string(self)?)
    }
}
